### PR TITLE
Fix arity of context.abort for AIO RPCs

### LINKED ISSUE #2065: grpc: trailing metadata can't be passed when aborting an AIO server RPC
**Describe your environment**

Python: 3.12.0
`opentelemetry-distro[otlp]==0.42b0`
`opentelemetry-instrumentation-grpc==0.42b0`
`grpcio==1.59.2`

**Steps to reproduce**
Full replication can be found [here](https://github.com/cookiefission/testbench/tree/main/otel/python-grpc-aio-abort-issue).

`server.py`:

```python
import asyncio
import logging

import grpc
import helloworld_pb2
import helloworld_pb2_grpc


class Greeter(helloworld_pb2_grpc.GreeterServicer):
    async def SayHello(
        self,
        request: helloworld_pb2.HelloRequest,
        context: grpc.aio.ServicerContext,
    ) -> helloworld_pb2.HelloReply:
        metadata = (
            ("this", "should"),
            ("work", "nicely")
        )
        await context.abort(code=grpc.StatusCode.ABORTED, details="This is the actual error message", trailing_metadata=metadata)


async def serve() -> None:
    server = grpc.aio.server()
    helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
    listen_addr = "[::]:50051"
    server.add_insecure_port(listen_addr)
    logging.info("Starting server on %s", listen_addr)
    await server.start()
    await server.wait_for_termination()


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(serve())
```
This server is just adopted from [the grpc examples](https://github.com/grpc/grpc/blob/master/examples/python/helloworld/async_greeter_server.py). The helloworld protobuf artefacts can be found [there](https://github.com/grpc/grpc/blob/master/examples/python/helloworld/helloworld_pb2_grpc.py).

For a client, the [`async_greeter_client.py`](https://github.com/grpc/grpc/blob/master/examples/python/helloworld/async_greeter_client.py) from grpc examples can be used.

This only happens when `trailing_metadata` is passed as an argument to `context.abort`. 

**What is the expected behavior?**

Running the server without opentelemetry, the (expected) error from running the client is:

```
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.ABORTED
        details = "This is the actual error message"
        debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"This is the actual error message", grpc_status:10, created_time:"2023-11-16T15:09:03.552303+00:00"}"
>
```

When running the server under `opentelemetry-instrument`, I expect the same output as when it is run without OTEL. 

**What is the actual behavior?**

Instead, an exception is still raised but the underlying code, details, and trailing_metadata are missing. This makes any nuanced error handling on the client side impossible. 

```
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Unexpected <class 'TypeError'>: _OpenTelemetryServicerContext.abort() got an unexpected keyword argument 'trailing_metadata'"
        debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2023-11-16T15:30:02.017999+00:00", grpc_status:2, grpc_message:"Unexpected <class \'TypeError\'>: _OpenTelemetryServicerContext.abort() got an unexpected keyword argument \'trailing_metadata\'"}"
>
```

**Additional context**

The `abort` method for [`grpc.ServicerContext`](https://grpc.github.io/grpc/python/grpc.html#grpc.ServicerContext.abort) and [`grpc.aio.ServicerContext`](https://grpc.github.io/grpc/python/grpc_asyncio.html#grpc.aio.ServicerContext.abort) is different. It's unclear why. The AIO instrumentation [re-uses](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_aio_server.py#L19) the [`_OpenTelemetryServicerContext`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py#L68) from the non-async server instrumentation which is how this bug snuck in, despite there being a test in place for the abort.

### PR DESCRIPTION
**Pull request by cookiefission**
# Description

This fixes an issue with using `context.abort` in a `grpc.aio` RPC servicer. In cases where the `trailing_metadata` argument is passed, the `abort` method would previous throw a `TypeError` because the `grpc.ServicerContext.abort` method didn't take `trailing_metadata`; the `grpc.aio.ServicerContext` does take that arg so should be matched. It's also true that `context.abort(code)` is valid for a `grpc.aio.ServicerContext` but not a `grpc.ServicerContext` where `details` is a required argument. 

Fixes #2065 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Unit tests
- [ ] Development version used in replication repo

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_aio_server.py
Comment (cookiefission) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_aio_server.py:27]: Yup, that makes sense. I'll give it a try today.

**Review by aabmass — APPROVED** [review]
Thanks for the PR! LGTM

**Comment by cookiefission** [comment]
Thanks for approving @aabmass . Is there anything left for me to do before it can be merged?

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_aio_server.py
Comment (FilipNikolovski) [inline instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_aio_server.py:53]: We should set the grpc status codes to adhere to the otel spec. There is already an open issue (and a PR) regarding this: https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1756

I think we should also make the changes here as well, to properly set the status codes.

Comment (cookiefission) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_aio_server.py:53]: Those changes haven't landed yet so I'm a little reluctant to add them here. More than happy to take a steer from a maintainer but I'd rather keep the scope of this bugfix small.

File: CHANGELOG.md
Comment (cookiefission) [reply CHANGELOG.md:16]: Done

**Pull request merged**
